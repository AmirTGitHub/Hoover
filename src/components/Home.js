import React, { Component, Fragment } from 'react';
import Header from './Header'
import RoomController from './RoomController'
import HooverController from './HooverController'
import Room from './Room'


export default class Home extends Component {
    state = {
        column: 0,
        row: 0,
        disabled: false,
        room: [],
        dirtPosition: 0,
        hooverPosition: 0,
        display: false,
    }

    handleColumn = event => {
        const column = parseInt(Math.abs(event.target.value), 10)
        this.setState({
            column: column
        })
    }
    handleRow = event => {
        const row = parseInt(Math.abs(event.target.value), 10)
        this.setState({
            row: row,
            hooverPosition: row
        })
    }

    roomCreator = () => {
        const row = this.state.row
        const column = this.state.column
        this.setState({
            disabled: true
        })
        let room = []
        for (let i = 0; i < row * column; i++) {
            room.push(i)
        }
        this.setState({
            room: room,
            display: true
        })
    }
    getDirtPosition = event => {
        let dirt = parseInt(Math.abs(event.target.value), 10)
        const row = this.state.row

        if (dirt === row) {
            alert('please select another cell')

        }
        this.setState({
            dirtPosition: dirt
        })
    }
    goLeft = () => {
        const hoover = this.state.hooverPosition
        this.setState({
            hooverPosition: hoover - 1
        })
        this.cleaning()
    }
    goDown = () => {
        const hoover = this.state.hooverPosition
        const row = this.state.row
        this.setState({
            hooverPosition: hoover + row
        })
        this.cleaning()
    }
    goRight = () => {
        const hoover = this.state.hooverPosition
        this.setState({
            hooverPosition: hoover + 1
        })
        this.cleaning()
    }
    goUp = () => {
        const hoover = this.state.hooverPosition
        const row = this.state.row

        this.setState({
            hooverPosition: hoover - row
        })
        this.cleaning()
    }

    cleaning = () => {
        const hoover = this.state.hooverPosition
        const dirt = this.state.dirtPosition
        if (hoover === dirt) {
            this.setState({
                dirtPosition: 0,
                hooverPosition: this.state.row
            })
        }

    }

    render() {
        return (
            <Fragment >
                <Header headerTitle='Hoover app' />
                {/* with this component I will create the room*/}
                <RoomController
                    createRoom={this.roomCreator}
                    getRow={this.handleRow}
                    getColumn={this.handleColumn}
                    disabled={this.state.disabled}
                    display={this.state.display}
                    maxNumber={this.state.row * this.state.column}
                    getDirt={this.getDirtPosition}
                    dirt={this.state.dirtPosition}
                />

                <hr />
                {/* when the room creator button clicked the room and the hover controller will be appear*/}
                {this.state.display ?
                    <div className='container' >
                        {/* this is the actual room*/}
                        <Room grid={this.state.room}
                            width={this.state.row}
                            dirt={this.state.dirtPosition}
                            hoover={this.state.hooverPosition}
                        />
                        {/* this is the hoover controller*/}
                        <HooverController
                            goLeft={this.goLeft}
                            goDown={this.goDown}
                            goRight={this.goRight}
                            goUp={this.goUp}
                        />
                    </div>
                    : null}
            </Fragment>
        )
    };
}