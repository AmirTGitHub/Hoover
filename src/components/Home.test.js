import Home from './Home'

const selector = (id) => `[data-test="${id}"]`

const firstState = {
    column: 0,
    row: 0,
    disabled: false,
    room: [],
    dirtPosition: 0,
    hooverPosition: 0,
    display: false
}

it('render the whole app', () => {
    const wrapper = render(
        <Home />
    )
    expect(wrapper).toMatchSnapshot()
})


it('check to have empty state at the beginning', () => {
    const wrapper = mount(<Home />)
    const emptyState = wrapper.state()
    expect(emptyState).toMatchObject(firstState)
})


const newState = {
    column: 10,
    row: 10,
    disabled: true,
    room: [...new Array(100)],
    dirtPosition: 48,
    hooverPosition: 10,
    display: true
}

it('render the room and the hoover controller when the display in state changed', () => {
    const wrapper = mount(<Home />)
    expect(wrapper).toMatchSnapshot()
    wrapper.setState({ ...newState })
    expect(wrapper).toMatchSnapshot()

})