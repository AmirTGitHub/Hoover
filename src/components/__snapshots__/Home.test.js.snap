// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the room and the hoover controller when the display in state changed 1`] = `
Array [
  <div
    className="header"
>
    <h1
        className="header--title"
    >
        Hoover app
    </h1>
</div>,
  <div>
    <div
        className="centre"
    >
        <form
            style={
                Object {
                    "padding": 50,
                  }
            }
        >
            <h3>
                Row
            </h3>
            <label />
            <input
                data-test="input"
                disabled={false}
                name="row"
                onChange={[Function]}
                placeholder="number"
                type="number"
            />
        </form>
        <form
            style={
                Object {
                    "padding": 50,
                  }
            }
        >
            <h3>
                Column
            </h3>
            <label />
            <input
                data-test="input"
                disabled={false}
                name="column"
                onChange={[Function]}
                placeholder="number"
                type="number"
            />
        </form>
    </div>
    <div
        className="centre"
    >
        <button
            className="btn"
            onClick={[Function]}
        >
             Create the room 
        </button>
    </div>
</div>,
  <hr />,
]
`;

exports[`render the room and the hoover controller when the display in state changed 2`] = `
Array [
  <div
    className="header"
>
    <h1
        className="header--title"
    >
        Hoover app
    </h1>
</div>,
  <div>
    <div
        className="centre"
    >
        <form
            style={
                Object {
                    "padding": 50,
                  }
            }
        >
            <h3>
                Row
            </h3>
            <label />
            <input
                data-test="input"
                disabled={true}
                name="row"
                onChange={[Function]}
                placeholder="number"
                type="number"
            />
        </form>
        <form
            style={
                Object {
                    "padding": 50,
                  }
            }
        >
            <h3>
                Column
            </h3>
            <label />
            <input
                data-test="input"
                disabled={true}
                name="column"
                onChange={[Function]}
                placeholder="number"
                type="number"
            />
        </form>
    </div>
    <div
        className="centre"
    >
        <form
            style={
                Object {
                    "padding": 50,
                  }
            }
        >
            <h3>
                Dirt position
            </h3>
            <label>
                please write a number between 0 and 100
            </label>
            <input
                data-test="input"
                name="dirt"
                onChange={[Function]}
                placeholder="number"
                type="number"
            />
        </form>
        <button
            className="btn"
            onClick={[Function]}
        >
             Create the room 
        </button>
    </div>
</div>,
  <hr />,
  <div
    className="container"
>
    <div
        className="roomFloor"
        style={
            Object {
                "width": "510px",
              }
        }
    >
        <div
            className="cell"
            id={1}
        />
        <div
            className="cell"
            id={2}
        />
        <div
            className="cell"
            id={3}
        />
        <div
            className="cell"
            id={4}
        />
        <div
            className="cell"
            id={5}
        />
        <div
            className="cell"
            id={6}
        />
        <div
            className="cell"
            id={7}
        />
        <div
            className="cell"
            id={8}
        />
        <div
            className="cell"
            id={9}
        />
        <img
            alt="a robotic hoover"
            className="cell"
            src="Hoover_01png"
        />
        <div
            className="cell"
            id={11}
        />
        <div
            className="cell"
            id={12}
        />
        <div
            className="cell"
            id={13}
        />
        <div
            className="cell"
            id={14}
        />
        <div
            className="cell"
            id={15}
        />
        <div
            className="cell"
            id={16}
        />
        <div
            className="cell"
            id={17}
        />
        <div
            className="cell"
            id={18}
        />
        <div
            className="cell"
            id={19}
        />
        <div
            className="cell"
            id={20}
        />
        <div
            className="cell"
            id={21}
        />
        <div
            className="cell"
            id={22}
        />
        <div
            className="cell"
            id={23}
        />
        <div
            className="cell"
            id={24}
        />
        <div
            className="cell"
            id={25}
        />
        <div
            className="cell"
            id={26}
        />
        <div
            className="cell"
            id={27}
        />
        <div
            className="cell"
            id={28}
        />
        <div
            className="cell"
            id={29}
        />
        <div
            className="cell"
            id={30}
        />
        <div
            className="cell"
            id={31}
        />
        <div
            className="cell"
            id={32}
        />
        <div
            className="cell"
            id={33}
        />
        <div
            className="cell"
            id={34}
        />
        <div
            className="cell"
            id={35}
        />
        <div
            className="cell"
            id={36}
        />
        <div
            className="cell"
            id={37}
        />
        <div
            className="cell"
            id={38}
        />
        <div
            className="cell"
            id={39}
        />
        <div
            className="cell"
            id={40}
        />
        <div
            className="cell"
            id={41}
        />
        <div
            className="cell"
            id={42}
        />
        <div
            className="cell"
            id={43}
        />
        <div
            className="cell"
            id={44}
        />
        <div
            className="cell"
            id={45}
        />
        <div
            className="cell"
            id={46}
        />
        <div
            className="cell"
            id={47}
        />
        <img
            alt="an imaginary dirt"
            className="cell"
            src="Dirt_01.png"
        />
        <div
            className="cell"
            id={49}
        />
        <div
            className="cell"
            id={50}
        />
        <div
            className="cell"
            id={51}
        />
        <div
            className="cell"
            id={52}
        />
        <div
            className="cell"
            id={53}
        />
        <div
            className="cell"
            id={54}
        />
        <div
            className="cell"
            id={55}
        />
        <div
            className="cell"
            id={56}
        />
        <div
            className="cell"
            id={57}
        />
        <div
            className="cell"
            id={58}
        />
        <div
            className="cell"
            id={59}
        />
        <div
            className="cell"
            id={60}
        />
        <div
            className="cell"
            id={61}
        />
        <div
            className="cell"
            id={62}
        />
        <div
            className="cell"
            id={63}
        />
        <div
            className="cell"
            id={64}
        />
        <div
            className="cell"
            id={65}
        />
        <div
            className="cell"
            id={66}
        />
        <div
            className="cell"
            id={67}
        />
        <div
            className="cell"
            id={68}
        />
        <div
            className="cell"
            id={69}
        />
        <div
            className="cell"
            id={70}
        />
        <div
            className="cell"
            id={71}
        />
        <div
            className="cell"
            id={72}
        />
        <div
            className="cell"
            id={73}
        />
        <div
            className="cell"
            id={74}
        />
        <div
            className="cell"
            id={75}
        />
        <div
            className="cell"
            id={76}
        />
        <div
            className="cell"
            id={77}
        />
        <div
            className="cell"
            id={78}
        />
        <div
            className="cell"
            id={79}
        />
        <div
            className="cell"
            id={80}
        />
        <div
            className="cell"
            id={81}
        />
        <div
            className="cell"
            id={82}
        />
        <div
            className="cell"
            id={83}
        />
        <div
            className="cell"
            id={84}
        />
        <div
            className="cell"
            id={85}
        />
        <div
            className="cell"
            id={86}
        />
        <div
            className="cell"
            id={87}
        />
        <div
            className="cell"
            id={88}
        />
        <div
            className="cell"
            id={89}
        />
        <div
            className="cell"
            id={90}
        />
        <div
            className="cell"
            id={91}
        />
        <div
            className="cell"
            id={92}
        />
        <div
            className="cell"
            id={93}
        />
        <div
            className="cell"
            id={94}
        />
        <div
            className="cell"
            id={95}
        />
        <div
            className="cell"
            id={96}
        />
        <div
            className="cell"
            id={97}
        />
        <div
            className="cell"
            id={98}
        />
        <div
            className="cell"
            id={99}
        />
        <div
            className="cell"
            id={100}
        />
    </div>
    <div
        className="hooverController"
    >
        <h1>
            Hoover Controller
        </h1>
        <button
            className="up btn-controller"
            onClick={[Function]}
        >
            <img
                alt="arrow up"
                src="arrow.png"
            />
        </button>
        <div>
            <button
                className="btn-controller"
                onClick={[Function]}
            >
                <img
                    alt="arrow left"
                    src="arrow.png"
                />
            </button>
            <button
                className="right btn-controller"
                onClick={[Function]}
            >
                <img
                    alt="arrow right"
                    src="arrow.png"
                />
            </button>
        </div>
        <button
            className="down btn-controller"
            onClick={[Function]}
        >
            <img
                alt="arrow down"
                src="arrow.png"
            />
        </button>
    </div>
</div>,
]
`;

exports[`render the whole app 1`] = `
Array [
  <div
    class="header"
>
    <h1
        class="header--title"
    >
        Hoover app
    </h1>
</div>,
  <div>
    <div
        class="centre"
    >
        <form
            style="padding:50px"
        >
            <h3>
                Row
            </h3>
            <label>
                
            </label>
            <input
                data-test="input"
                name="row"
                placeholder="number"
                type="number"
            >
                
            </input>
        </form>
        <form
            style="padding:50px"
        >
            <h3>
                Column
            </h3>
            <label>
                
            </label>
            <input
                data-test="input"
                name="column"
                placeholder="number"
                type="number"
            >
                
            </input>
        </form>
    </div>
    <div
        class="centre"
    >
        <button
            class="btn"
        >
             Create the room 
        </button>
    </div>
</div>,
  <hr>
    
</hr>,
]
`;
